### PR TITLE
gh-112320: Implement on-trace confidence tracking for branches

### PR DESCRIPTION
**Pull request by gvanrossum**
- Issue: gh-112320.

### CODE REVIEW

**Comment by gvanrossum** [comment]
@markshannon Is this what you had in mind for confidence tracking? I just picked a random cut-off of about 1/3rd.

**Comment by gvanrossum** [comment]
@mdboom Could you check that I did everything right for adding a new field to the stats?

File: Python/optimizer.c [hunk 1]
```diff
@@ -543,11 +543,23 @@ translate_bytecode_to_trace(
                 int counter = instr[1].cache;
                 int bitcount = _Py_popcount32(counter);
                 int jump_likely = bitcount > 8;
+                if (jump_likely) {
+                    confidence *= bitcount / 16.0;
```
Comment (markshannon) [inline Python/optimizer.c]: Division is notoriously slow and beware of implicit casts to `double`, so can you make this `bitcount * (float)(1/16.0)`.

File: Python/optimizer.c [hunk 2]
```diff
@@ -543,11 +543,23 @@ translate_bytecode_to_trace(
...
+                }
+                else {
+                    confidence *= (16 - bitcount) / 16.0;
```
Comment (markshannon) [inline Python/optimizer.c]: Ditto

File: Python/optimizer.c [hunk 3]
```diff
@@ -543,11 +543,23 @@ translate_bytecode_to_trace(
...
+                }
+                // TODO: Tweak the confidence cutoff
+                if (confidence < 0.33) {
```
Comment (markshannon) [inline Python/optimizer.c]: 0.33 is fine for now, but can you make it a symbolic constant?

**Comment by markshannon** [comment]
Yes, this is what I had in mind.

We could also reduce the confidence for type checks, although it might not make a difference in practice unless we increase the max trace length considerably.

**Pull request merged**